### PR TITLE
Maximum linkage errors is really new linkage errors

### PR DESCRIPTION
**Pull request by elharo**
@suztomo

### CODE REVIEW

**Comment by elharo** [comment]
java.lang.AssertionError: Baseline BOM: com.google.cloud:libraries-bom:4.2.0
Newly introduced problems:
Class io.reactivex.internal.schedulers.NonBlockingThread is not found
Class reactor.core.scheduler.NonBlocking is not found
Class com.sun.jna.Native is not found
Class com.sun.jna.Structure is not found
Class com.sun.jna.platform.win32.WinNT$HANDLE is not found
Class com.sun.jna.ptr.IntByReference is not found
Class com.sun.jna.platform.win32.Kernel32 is not found
Class com.sun.jna.platform.win32.Win32Exception is not found
Class com.sun.jna.Platform is not found
Class com.sun.jna.platform.win32.WinDef$LPVOID is not found
Class com.sun.jna.Library is not found
Class com.sun.jna.LastErrorException is not found
Class com.sun.jna.platform.win32.WinDef$DWORD is not found
Class com.sun.jna.Memory is not found
Class com.sun.jna.Pointer is not found
Class com.sun.jna.win32.W32APIOptions is not found
Class com.sun.jna.win32.StdCallLibrary is not found
Class com.sun.jna.platform.win32.WinNT$SECURITY_DESCRIPTOR is not found
Class com.sun.jna.platform.win32.Advapi32 is not found
Class com.sun.jna.platform.win32.WinBase$SECURITY_ATTRIBUTES is not found

**Comment by elharo** [comment]
some more details: looks like something called blockhound is at fault:

java.lang.AssertionError: Baseline BOM: com.google.cloud:libraries-bom:4.2.0
Newly introduced problems:
Class io.reactivex.internal.schedulers.NonBlockingThread is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.integration.RxJava2Integration}]
Class reactor.core.scheduler.NonBlocking is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.integration.ReactorIntegration}]
Class com.sun.jna.Native is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.Structure is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.platform.win32.WinNT$HANDLE is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.ptr.IntByReference is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.platform.win32.Kernel32 is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.platform.win32.Win32Exception is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.Platform is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.platform.win32.WinDef$LPVOID is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.Library is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.LastErrorException is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.platform.win32.WinDef$DWORD is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.Memory is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.Pointer is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.win32.W32APIOptions is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.win32.StdCallLibrary is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.platform.win32.WinNT$SECURITY_DESCRIPTOR is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.platform.win32.Advapi32 is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]
Class com.sun.jna.platform.win32.WinBase$SECURITY_ATTRIBUTES is not found referenced from [ClassFile{jar=/home/elharo/.m2/repository/io/projectreactor/tools/blockhound/1.0.1.RELEASE/blockhound-1.0.1.RELEASE.jar, className=reactor.blockhound.shaded.net.bytebuddy.agent.VirtualMachine}]

**Comment by suztomo** [comment]
Try generating the dashboard in your Git branch. It gives dependency tree to explain why the errors happen.

**Comment by elharo** [comment]
Problem is this release of gRPC pulls io.projectreactor.tools:blockhound:1.0.1.RELEASE 
and that has many new linkage errors.

**Comment by suztomo** [comment]
Memo:
Why does `io.projectreactor.tools:blockhound:1.0.1.RELEASE` not have its depending classes?

- `com.sun.jna` seems platform-dependent classes. What are they?
- Missing `io.reactivex` and `reactor.core.scheduler` look genuine linkage errors that are usually caused by picking up wrong version of a library.

**Comment by suztomo** [comment]
From dashboard:

> The following paths contain blockhound-1.0.1.RELEASE.jar:
> 
> io.grpc:grpc-netty:1.28.0 (compile) / io.netty:netty-codec-http2:4.1.45.Final (compile) / io.netty:netty-common:4.1.45.Final (compile) / io.projectreactor.tools:blockhound:1.0.1.RELEASE (compile, optional)
> io.grpc:grpc-netty:1.28.0 (compile) / io.netty:netty-codec-http2:4.1.45.Final (compile) / io.netty:netty-common:4.1.45.Final (compile) / io.projectreactor.tools:blockhound:1.0.1.RELEASE (compile, optional)
> io.grpc:grpc-netty:1.28.0 (compile) / io.netty:netty-handler-proxy:4.1.45.Final (compile) / io.netty:netty-common:4.1.45.Final (compile) / io.projectreactor.tools:blockhound:1.0.1.RELEASE (compile, optional)
> io.grpc:grpc-netty:1.28.0 (compile) / io.netty:netty-codec-http2:4.1.45.Final (compile) / io.netty:netty-buffer:4.1.45.Final (compile) / io.netty:netty-common:4.1.45.Final (compile) / io.projectreactor.tools:blockhound:1.0.1.RELEASE (compile, optional)

**Comment by suztomo** [comment]
Netty-common has `NettyBlockHoundIntegration` class. BlockHound https://github.com/reactor/BlockHound is a java agent that detects blocking Java API; a quite interesting solution to ensure Netty usage is non-blocking.

Those who want to use BlockHound for Netty touch NettyBlockHoundIntegration class.

Netty declares BlockHound as optional:
```
    <dependency>
      <groupId>io.projectreactor.tools</groupId>
      <artifactId>blockhound</artifactId>
      <version>1.0.1.RELEASE</version>
      <scope>compile</scope>
      <optional>true</optional>
    </dependency>
```

BlockHound is not declaring dependencies in its pom.xml. Created an issue there https://github.com/reactor/BlockHound/issues/99

**Comment by bsideup** [comment]
@suztomo BlockHound either shades it's dependencies (like ByteBuddy that has optional JNA dependency that we don't include) or uses `compileOnly` for _optional_ integrations that are triggered only when class is available.

JFYI adding and installing BlockHound will trigger Java's SPI mechanism (`ServiceLoader`) and load available integrations (including `NettyBlockHoundIntegration`).

**Comment by elharo** [comment]
1. Shading loses the optionality of the dependency.

2. Blockhound doesn't list any dependencies at all in its pom.xml: https://repo1.maven.org/maven2/io/projectreactor/tools/blockhound/1.0.2.RELEASE/

compileOnly is a gradle thing. Maven doesn't really have a precise equivalent of that.

**Comment by bsideup** [comment]
@elharo 

> Shading loses the optionality of the dependency.

Shading removes any trail of the dependency and that's exactly what we wanted - to not depend on something in any way.

> Blockhound doesn't list any dependencies at all in its pom.xml:

It does not indeed. We _might_ list optional dependencies like RxJava and Reactor, but it won't help with ByteBuddy's JNA for example. Could you please help me understanding what your tool is doing and why it requires us to list our `compileOnly` dependencies? Since, from the linking PoV (given JVM's dynamic nature) we are fine.

**Comment by elharo** [comment]
The tool is analyzing the class files and looking for linkage errors; i.e. places where code is referencing a method or class that isn't present: 

https://github.com/GoogleCloudPlatform/cloud-opensource-java/tree/master/linkage-monitor

What we're actually looking at here is gRPC which depends on Netty which depends on ByteBuddy (and didn't used to). 

The shading in ByteBuddy is incomplete. There are references left to classes that are not present. Possibly they're unreachable or guarded by other code, but if so we need to verify that carefully and figure out how to handle it. It sounds like something in the middle might need to exclude ByteBuddy or mark it optional or something, but we need to figure out what's really going on here first.

**Comment by bsideup** [comment]
@elharo 
> gRPC which depends on Netty which depends on ByteBuddy 

on BlockHound, which shades ByteBuddy.

> The shading in ByteBuddy is incomplete. There are references left to classes that are not present. 

In BlockHound you mean? Shading does not guarantee that all references will be present. it is complete, just some classes (like JNA) will only be used if they are available on classpath.
Other classes (like RxJava's) are not supposed to be shaded and also not used unless available on classpath.

I would also like to mention that BlockHound works. The linkage checker seems to perform some _static_ analysis and gives false negatives. I think we could add JNA as an optional dependency, just to make it happier. But I don't think RxJava and Reactor should be listed there, see my answer in BlockHound's ticket:
https://github.com/reactor/BlockHound/issues/99#issuecomment-599529306

**Comment by elharo** [comment]
You're right. I meant in BlockHound.

File: boms/cloud-oss-bom/pom.xml [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
   <groupId>com.google.cloud</groupId>
   <artifactId>libraries-bom</artifactId>
-  <version>4.2.1-SNAPSHOT</version>
+  <version>4.3.0-SNAPSHOT</version>
```
Comment (suztomo) [inline boms/cloud-oss-bom/pom.xml]: @elharo I see the gRPC version change has been removed from this PR. You may want to exclude this version change too.

Comment (elharo) [reply boms/cloud-oss-bom/pom.xml]: I'm assuming the next version change will pick up a minor, not patch, release but if not we can always change this again.

**Pull request merged**